templates/ingress-gateways-deployment.yaml

```diff
@@ -0,0 +1,376 @@
+                {{- if and (eq $source "Service") (not $serviceEnabled) }}{{ fail "if ingressGateways .wanAddress.source=Service then ingressGateways .service.enabled must be set to true in either the defaults or in the gateway definition" }}{{ end }}
+                {{- if or (eq $source "NodeIP") (and (eq $source "Service") (eq $serviceType "NodePort")) }}
+                WAN_ADDR="${HOST_IP}"
+                {{- else if eq $source "NodeName" }}
+                WAN_ADDR="${NODE_NAME}"
+                {{- else if and (eq $source "Service") (or (eq $serviceType "ClusterIP") (eq $serviceType "LoadBalancer")) }}
+                consul-k8s service-address \
+                  -k8s-namespace={{ $root.Release.Namespace }} \
+                  -name={{ .name }} \
+                  -output-file=/tmp/address.txt
+                WAN_ADDR="$(cat /tmp/address.txt)"
+                {{- else if eq $source "Static" }}
+                {{- if eq (default $defaults.wanAddress.static $wanAddress.static) "" }}{{ fail "if ingressGateways .wanAddress.source=Static then ingressGateways .wanAddress.static cannot be empty and must be set in either the defaults or in the gateway definition" }}{{ end }}
+                WAN_ADDR="{{ (default $defaults.wanAddress.static $wanAddress.static) }}"
+                {{- else }}
+                {{- fail "currently set ingressGateway values for wanAddress.source and service.type are not supported" }}
+                {{- end }}
+
+                {{- if eq $source "Service" }}
+                {{- if eq $serviceType "NodePort" }}
+                {{- if not (default $defaults.service.nodePort $service.nodePort) }}{{ fail "if ingressGateways .wanAddress.source=Service and ingressGateways .service.type=NodePort, ingressGateways .service.nodePort must be set in either the defaults or in the gateway definition" }}{{ end }}
+                WAN_PORT="{{ (default $defaults.service.nodePort $service.nodePort) }}"
+                {{- else }}
+                WAN_PORT="{{ (default $defaults.service.port $service.port) }}"
+                {{- end }}
+                {{- else }}
+                WAN_PORT="{{ (default $defaults.wanAddress.port $wanAddress.port) }}"
+                {{- end }}
+
+                cat > /consul/service/service.hcl << EOF
+                service {
+                  kind = "ingress-gateway"
+                  name = "{{ .name }}"
+                  {{- if $root.Values.global.enableConsulNamespaces }}
+                  namespace = "{{ (default $defaults.consulNamespace .consulNamespace) }}"
+                  {{- end }}
+                  port = ${WAN_PORT}
+                  address = "${WAN_ADDR}"
+                  tagged_addresses {
+                    lan {
+                      address = "${POD_IP}"
+                      port = 8443
+                    }
+                    wan {
+                      address = "${WAN_ADDR}"
+                      port = ${WAN_PORT}
```

Comment:
todo: do we need to define a port? This is only used for DNS so then we'd only need the IP, however if they're using SRV records then maybe we'd use the port? But as we figured out this port can't actually be used for any ingress traffic, only for the healthcheck.

need to check if this port is returned in an SRV lookup. Otherwise I think we can remove it.

Reply:
Ignore. This is good!